redirect/src/main/java/com/adyen/checkout/redirect/RedirectUtil.kt

```diff
@@ -100,35 +77,99 @@ object RedirectUtil {
         }
     }
 
-    /**
-     * Creates the intent that will start the redirect.
-     * @param context Any context.
-     * @param uri The Uri to redirect to.
-     * @return And intent that targets either another app or a Web page.
-     */
-    @JvmStatic
-    fun createRedirectIntent(context: Context, uri: Uri): Intent {
-        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
-            createCustomTabIntent(context, uri)
-        } else {
-            if (determineResolveResult(context, uri).type == ResolveResult.Type.APPLICATION) {
-                Intent(Intent.ACTION_VIEW, uri)
-            } else {
-                createCustomTabIntent(context, uri)
-            }
+    @Suppress("ReturnCount")
+    fun launchUriRedirect(context: Context, uri: Uri) {
+        if (launchNative(context, uri)) return
+        if (launchWithCustomTabs(context, uri)) return
+        if (launchBrowser(context, uri)) return
+        throw ComponentException("Redirect to app failed.")
+    }
+
+    private fun launchNative(context: Context, uri: Uri): Boolean {
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) launchNativeApi30(context, uri)
+        else launchNativeBeforeApi30(context, uri)
+    }
+
+    private fun launchNativeBeforeApi30(context: Context, uri: Uri): Boolean {
+        val pm = context.packageManager
+
+        // Get all Apps that resolve a generic url
+        val browserActivityIntent = Intent()
+            .setAction(Intent.ACTION_VIEW)
+            .addCategory(Intent.CATEGORY_BROWSABLE)
+            .setData(Uri.fromParts("http", "", null))
+        val genericResolvedSet = pm.queryIntentActivities(browserActivityIntent, 0).map {
+            it.resolvePackageName
+        }.toSet()
+
+        // Get all apps that resolve the specific Url
+        val specializedActivityIntent = Intent(Intent.ACTION_VIEW, uri)
+            .addCategory(Intent.CATEGORY_BROWSABLE)
+        val resolvedSpecializedSet = pm.queryIntentActivities(specializedActivityIntent, 0).map {
+            it.resolvePackageName
+        }.toMutableSet()
+
+        // Keep only the Urls that resolve the specific, but not the generic
+        // urls.
+        resolvedSpecializedSet.removeAll(genericResolvedSet)
+
+        // If the list is empty, no native app handlers were found.
+        if (resolvedSpecializedSet.isEmpty()) {
+            return false
+        }
+
+        // We found native handlers. Launch the Intent.
+        specializedActivityIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+        context.startActivity(specializedActivityIntent)
+        return true
+    }
+
+    private fun launchNativeApi30(context: Context, uri: Uri?): Boolean {
```

Comment:
Minor: you can add `@RequiresApi(Build.VERSION_CODES.R)` here to remove the warning.